src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -647,6 +701,13 @@ private void finish(@Nonnull Result r, @CheckForNull Throwable t) {
         } catch (Exception x) {
             LOGGER.log(Level.WARNING, "failed to save " + this + " or perform log rotation", x);
         }
+        if (culprits == null) {
+            HashSet<String> tempCulpritIds = new HashSet<>();
+            for (User u : getCulprits()) {
+                tempCulpritIds.add(u.getId());
+            }
+            culprits = ImmutableSortedSet.copyOf(tempCulpritIds);
```

Comment:
So we are computing this only at build completion. What about a `mail` step in a `post` `failure` block? When the `result` is still `null`? It is not obvious that the culprits logic will work at all in any practical situation. Needs some integration tests.